### ISSUE TITLE
Compiler features detection and discovery

### ISSUE THREAD

**Issue by KOLANICH** [SanderMertens/bake#44]
Different compilers, linkers and other tools of toolchains have different features and different flags controlling the same features. So the following facilities are needed:
* feature abstraction layer
* feature detection
* mapping flags to features, configured in config files in known directory
* optionally-enabled compiler features - a feature is enabled if it is supported by a tool
* features presets - a set of features enabled for a tool, configured in config files in known directory

**Comment by SanderMertens** [comment]
I agree. The current driver implementations for gcc/clang and msvc are tied to the platform bake is running on, which is inflexible. A feature/toolchain abstraction layer would address that.

Do you have any features in mind? It will be easier to design such a layer knowing which kinds of features it should address.

**Comment by KOLANICH** [comment]
>Do you have any features in mind?

Mostly [different hardening options](https://gitlab.com/KOLANICH/SQLiteCpp/blob/fixes_by_KOLANICH/cmake/Hardening.cmake), language versions, exceptions and RTTI toggles.

**Comment by SanderMertens** [comment]
Thanks, I see what you mean now! That would be very useful indeed. Currently bake lets you configure a few things, which are forwarded to the drivers which can then do what they want:

https://github.com/SanderMertens/bake#configuring-bake

This is not very sophisticated, and the list of options is still very limited. They basically translate into `if..then` statements in the driver implementations (for gcc, clang, msvc) to add or not add specific compiler options. Being able to detect available options would be nice.

It does sound like quite a bit of work, so it may take a while to get this implemented. First priority is to get an abstraction in place for running toolchains independent on platform (currently in progress). Once that's done, it should be easier to see how this can be integrated with that.
